crates/ruff_python_formatter/src/expression/number.rs

```diff
@@ -0,0 +1,191 @@
+
+        let normalized = normalize_floating_number(content.trim_end_matches(['j', 'J']));
+
+        match normalized {
+            Cow::Borrowed(_) => {
+                source_text_slice(range.sub_end(TextSize::from(1)), ContainsNewlines::No).fmt(f)?;
+            }
+            Cow::Owned(normalized) => {
+                dynamic_text(&normalized, Some(range.start())).fmt(f)?;
+            }
+        }
+
+        text("j").fmt(f)
+    }
+}
+
+/// Returns the normalized integer string.
+fn normalize_integer(input: &str) -> Cow<str> {
+    // The normalized string if `input` is not yet normalized.
+    // `output` must remain empty if `input` is already normalized.
+    let mut output = String::new();
+    // Tracks the last index of `input` that has been written to `output`.
+    // If `last_index` is `0` at the end, then the input is already normalized and can be returned as is.
+    let mut last_index = 0;
+
+    let mut is_hex = false;
+
+    let mut chars = input.char_indices();
+
+    if let Some((_, '0')) = chars.next() {
+        if let Some((index, c)) = chars.next() {
+            is_hex = matches!(c, 'x' | 'X');
+            if matches!(c, 'B' | 'O' | 'X') {
+                // Lowercase the prefix.
+                output.push('0');
+                output.push(c.to_ascii_lowercase());
+                last_index = index + c.len_utf8();
+            }
+        }
+    }
+
+    // Skip the rest if `input` is not a hexinteger because there are only digits.
+    if is_hex {
+        for (index, c) in chars {
+            if matches!(c, 'a'..='f') {
+                // Uppercase hexdigits.
+                output.push_str(&input[last_index..index]);
+                output.push(c.to_ascii_uppercase());
+                last_index = index + c.len_utf8();
+            }
+        }
+    }
+
+    if last_index == 0 {
+        Cow::Borrowed(input)
+    } else {
+        output.push_str(&input[last_index..]);
+        Cow::Owned(output)
+    }
+}
+
+/// Returns the normalized floating number string.
+fn normalize_floating_number(input: &str) -> Cow<str> {
+    // The normalized string if `input` is not yet normalized.
+    // `output` must remain empty if `input` is already normalized.
+    let mut output = String::new();
+    // Tracks the last index of `input` that has been written to `output`.
+    // If `last_index` is `0` at the end, then the input is already normalized and can be returned as is.
+    let mut last_index = 0;
+
+    let mut has_exponent = false;
+
+    let mut chars = input.char_indices();
+
+    if let Some((index, '.')) = chars.next() {
+        // Add a leading `0` if `input` starts with `.`.
+        output.push('0');
+        output.push('.');
+        last_index = index + '.'.len_utf8();
+    }
+
+    for (index, c) in chars {
+        if matches!(c, 'e' | 'E') {
+            has_exponent = true;
+            if input.as_bytes().get(index - 1).copied() == Some(b'.') {
```

Comment:
I believe this could, at least in theory, result in indexing between two character boundaries if the text starts with `[e10]`. I don't think this is a problem in practice because it's impossible that the preceding character is a unicode character, because the number would then be part of an identifier. I still recommend fixing this just to be sure (e.g. by assigning `is_float on line 147)